packages/bug-detectors/internal/prototype-pollution.ts

```diff
@@ -0,0 +1,565 @@
+	 * However, it also might result in false positives. Use `addExcludedExactMatch` to exclude specific values from
+	 * being reported as Prototype Pollution findings.
+	 */
+	instrumentAssignmentsAndVariableDeclarations(): PrototypePollutionConfig {
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		if (global.options.dryRun) {
+			console.error(
+				"[Prototype Pollution Configuration] The configuration option " +
+					"instrumentAssignmentsAndVariableDeclarations() is not supported in dry run mode.\n" +
+					"  Either disable dry-run mode or remove this option from custom hooks.\n" +
+					"  Jazzer.js initial arguments:"
+			);
+			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+			// @ts-ignore
+			console.error(global.options);
+			// We do not accept conflicting configuration options: abort.
+			process.exit(1);
+		}
+		this._instrument = true;
+		return this;
+	}
+
+	getExcludedExactMatches(): string[] {
+		return this.excludedExactMatches;
+	}
+
+	getInstrumentAssignmentsAndVariableDeclarations(): boolean {
+		return this._instrument;
+	}
+}
+
+const config: PrototypePollutionConfig = new PrototypePollutionConfig();
+
+// Add this bug detector's config to the global config map.
+bugDetectorConfigurations.set("prototype-pollution", config);
+
+interface PrototypePollution {
+	getProtoSnapshot: typeof getProtoSnapshot;
+	detectPrototypePollution: typeof detectPrototypePollution;
+	protoSnapshotsEqual: typeof protoSnapshotsEqual;
+}
+
+declare global {
+	// eslint-disable-next-line no-var
+	var PrototypePollution: PrototypePollution;
+}
+
+// Make these functions available to instrumentation plugins and the user via the global object.
+globalThis.PrototypePollution = {
+	getProtoSnapshot: getProtoSnapshot,
+	detectPrototypePollution: detectPrototypePollution,
+	protoSnapshotsEqual: protoSnapshotsEqual,
+};
+
+// Keep track of statements and expressions that should not be instrumented.
+// This is necessary to avoid infinite recursion when instrumenting code.
+class InstrumentationGuard {
+	private map: Map<TrackedExpressions, Set<string>> = new Map();
+
+	add(expressionType: TrackedExpressions, value: unknown) {
+		if (!this.map.has(expressionType)) {
+			this.map.set(expressionType, new Set());
+		}
+		this.map.get(expressionType)?.add(JSON.stringify(value));
+	}
+
+	has(expression: TrackedExpressions, value: unknown): boolean {
+		return (
+			(this.map.has(expression) &&
+				this.map.get(expression)?.has(JSON.stringify(value))) ??
+			false
+		);
+	}
+}
+
+enum TrackedExpressions {
+	ASSIGNMENT_EXPRESSION,
+	VARIABLE_DECLARATION,
+}
+
+const instrumentationGuard = new InstrumentationGuard();
+
+registerInstrumentationPlugin((): PluginTarget => {
+	function getIdentifierFromAssignmentExpression(
+		expr: AssignmentExpression
+	): Identifier | undefined {
+		if (types.isIdentifier(expr.left)) {
+			return expr.left;
+		}
+
+		const skipMemberExpressions = (
```

Comment:
The body doesn't seem to close over anything `expr` is shadowed. If that's true, making it a top-level function would improve readability.

Reply:
I think I wanted to tell that it's only used inside that one function, but it's more confusing that way. Extracted to a stand-alone function.